PluginsAndFeatures/azure-toolkit-for-intellij/azure-sdk-reference-book/src/main/java/com/microsoft/azure/toolkit/intellij/azuresdk/service/WorkspaceTaggingService.java

```diff
@@ -0,0 +1,89 @@
+ */
+
+package com.microsoft.azure.toolkit.intellij.azuresdk.service;
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.MappingIterator;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.json.JsonMapper;
+import com.fasterxml.jackson.dataformat.csv.CsvMapper;
+import com.fasterxml.jackson.dataformat.csv.CsvSchema;
+import com.microsoft.azure.toolkit.intellij.azuresdk.model.AzureJavaSdkEntity;
+import com.microsoft.azure.toolkit.intellij.azuresdk.model.WorkspaceTagEntity;
+import com.microsoft.azure.toolkit.lib.common.cache.Cacheable;
+import com.microsoft.azure.toolkit.lib.common.cache.Preload;
+import org.apache.commons.lang3.ObjectUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.List;
+
+public class WorkspaceTaggingService {
+    private static final String SDK_METADATA_URL = "https://raw.githubusercontent.com/Azure/azure-sdk/master/_data/releases/latest/java-packages.csv";
+
+    private static final ObjectMapper CSV_MAPPER = new CsvMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+    private static final ObjectMapper JSON_MAPPER = new JsonMapper().configure(JsonParser.Feature.ALLOW_COMMENTS, true);
+    private static final String WORKSPACE_TAG_JSON = "/workspaceTag.json";
+
+    @Nullable
+    public static String getWorkspaceTag(String groupId, final String artifactId) {
+        if (StringUtils.isAnyEmpty(groupId, artifactId)) {
+            return null;
+        }
+        return ObjectUtils.firstNonNull(getAzureDependencyTag(groupId, artifactId), getExternalDependencyTag(groupId, artifactId));
+    }
+
+    private static String getAzureDependencyTag(final String groupId, final String artifactId) {
+        try {
+            return getAzureSDKEntities()
+                .stream()
+                .filter(entity -> StringUtils.isNotEmpty(entity.getType())
+                    && StringUtils.equalsIgnoreCase(entity.getGroupId(), groupId)
+                    && StringUtils.equalsIgnoreCase(entity.getPackageName(), artifactId))
+                .map(AzureJavaSdkEntity::getType)
+                .findFirst().orElse(null);
+        } catch (IOException e) {
+            // swallow exception for workspace tagging
+            return null;
+        }
+    }
+
+    private static String getExternalDependencyTag(final String groupId, final String artifactId) {
+        try {
+            return getWorkspaceTagEntities()
+                .stream()
+                .filter(entity -> (StringUtils.isEmpty(entity.getGroupId()) || StringUtils.equalsIgnoreCase(entity.getGroupId(), groupId))
+                    && (StringUtils.isEmpty(entity.getArtifactId()) || StringUtils.equalsIgnoreCase(entity.getArtifactId(), artifactId)))
```

Comment:
seems no need to do empty check.

Reply:
If no value group id/ artifact id, we will accept all values, for instance, for spring boot, we will just check the group id and no validation on artifactId